### ISSUE TITLE
Paint/presentation timestamps in performance APIs

### ISSUE THREAD

**Issue by noamr** [w3ctag/design-reviews#1013]
こんにちは TAG-さん!

I'm requesting an early TAG design review of Paint/presentation timestamps in performance APIs.

Aligning different paint-related timestamps in all the relevant performance APIs, by having one "interoperable" timestamp
and one coarsened platform-specific "VSync" (or "pixels on screen") timestamp.

  - Explainer¹: https://github.com/w3c/paint-timing/blob/main/presentation-timestamps.md
  - User research:
  - Security and Privacy self-review²: https://github.com/w3c/paint-timing/pull/107
  - GitHub repo: https://github.com/w3c/paint-timing/
  - Primary contacts:
      - @noamr, Google Inc, editor
      - @mmocny, Google Inc, editor
  - Organization/project driving the design: Google
  - Multi-stakeholder feedback³:
    - Chromium comments: Supportive (driving)
    - Mozilla comments: hhttps://github.com/mozilla/standards-positions/issues/1110
    - WebKit comments: https://github.com/WebKit/standards-positions/issues/426
    - {{...include feedback/review from developers, implementers, civil society, and others}}

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://www.w3.org/TR/design-principles/)
  - The group where the incubation/design work on this is being done (or is intended to be done in the future):
  - The group where standardization of this work is intended to be done ("unknown" if not known):
  - Existing major pieces of multi-implementer review or discussion of this design:
  - Major unresolved issues with or opposition to this design:
  - This work is being funded by:

You should also know that...

This has been discussed in the WebPerfWG in detail over the last 2 years.

**Comment by matatk** [comment]
Hi @noamr, thanks for this review request also. We have been discussing paint/presentation timestamps today too, and have a couple of question about the explainer. We would really like to see the (end-)user needs made explicit. We'd also like to know about the alternatives that were considered, and why they found to be unsuitable, again this helps us weigh up the trade-offs. Thank you in advance for any further info you can provide.

*Just in case it eases future reference: we put some [background info on explainers in the other thread](https://github.com/w3ctag/design-reviews/issues/1012#issuecomment-2594023351)*

**Comment by noamr** [comment]
Sorry, I overlooked this message from the summer.
I've updated the explainer with user value and considered alternatives.
Please reopen and take a look when convenient! (I don't have permissions to re-open)

**Comment by xiaochengh** [comment]
Hi @noamr, the TAG discussed it today and found some new concerns:

We are aware that the "presentation time" is not well defined, and an accurate definition might be out of the scope of the Web specs. However, the "paint time" as how it's currently defined is not perceivable by the end user. It's somewhere in the middle of the entire rendering pipeline, and there might be an arbitrary delay between this and when the pixel actually shows up on the screen. For this reason, we are unsure whether this "paint time" is a good enough performance indicator to add to the platform.

We are also wondering if new timings in the rendering pipeline could be used to perform side-channel attacks, e.g. exfiltrate canvas data to someone else.

Have you considered alternatives that are more interoperable and better match the page performance from the end users' perspective?

**Comment by noamr** [comment]
> Hi [@noamr](https://github.com/noamr), the TAG discussed it today and found some new concerns:
> 
> We are aware that the "presentation time" is not well defined, and an accurate definition might be out of the scope of the Web specs. However, the "paint time" as how it's currently defined is not perceivable by the end user. It's somewhere in the middle of the entire rendering pipeline, and there might be an arbitrary delay between this and when the pixel actually shows up on the screen. For this reason, we are unsure whether this "paint time" is a good enough performance indicator to add to the platform.

It is the latest interoperable timestamp. It gives the developer indications about what they did that could have contributed to a long paint. It is also consistent across different performance entry types (e.g. LCP and LoAF), which allows correlation.
It's a very useful timestamp.

Note that most timestamps in existing entries (e.g. resource timing) are not directly reflecting UX, but are used for *attribution* - finding the root cause for problematic user experience. Paint timing is one of those.

> 
> We are also wondering if new timings in the rendering pipeline could be used to perform side-channel attacks, e.g. exfiltrate canvas data to someone else.

Presentation times are heavily coarsened when not in a cross-origin isolated environment for that reason.
This is mentioned in the [privacy & security section of the explainer](https://github.com/w3c/paint-timing/blob/main/presentation-timestamps.md#06-do-the-features-in-your-specification-expose-information-about-the-underlying-platform-to-origins).
> 
> Have you considered alternatives that are more interoperable and better match the page performance from the end users' perspective?

This has been discussed for about 4 years, see https://github.com/w3c/paint-timing/issues/62.
So, yes, we have considered this, and this is the best alternative we could come up with.

**Comment by xiaochengh** [comment]
Hi @noamr, the TAG discussed it again at a Breakout today. Thanks for addressing our security and privacy concerns!

Regarding the usefulness, we would like to see more evidence. In particular, a study on when the new paint time is and is not a good performance indicator will help a lot. This will also help developers use it correctly.

The thread https://github.com/w3c/paint-timing/issues/62 didn't discuss much about the usefulness, but mainly focused on the technical soundness of the new paint time. And there are obvious scenarios where the new paint time might not be a good performance indicator. For example: image, video, gaming, video conference, ... (and anything with a significant graphics workload in general). We need a more thorough understanding here.

**Comment by mmocny** [comment]
I agree that https://github.com/w3c/paint-timing/issues/62 doesn't have adequate detail about why we arrived at this specific solution.

For better historical context, I would read through:
- https://github.com/w3c/largest-contentful-paint/issues/104
  - As far back as 2022 this was the [proposed solution agreed amongst multiple implementers](https://github.com/w3c/largest-contentful-paint/issues/104#issuecomment-1271823151)
  - Was attempted as an interop project in 2023 but was too nascent
  - Was finally done as part of interop 2025 (and is now already part of baseline)
- Far more discussion, and links to Perf WG meeting minutes linked from https://github.com/w3c/largest-contentful-paint/issues/91

---

But let me also try to summarize the status and value:

- `paintTime` is technically not a user perceivable timepoint, and technically only marks an internal scheduling timestamp.
- It is not quite as valuable as `presentationTime`, which indeed does better to capture the effects of graphics workload, and is more accurately representative of user experience.
- However, `presentationTime` is not interoperably available across all browsers, nor is it strictly better in every way.
  - Chromium has been using this timestamp for various metrics (such as FCP, LCP, Event Timing, etc) for years.  Mozilla have signalled a willingness to implement it.  Safari has thus far not signalled a willingness to do so.
- `paintTime` also offers a unique value relative to `presentationTime`: it marks the final time point of each animation frame _on the main thread_.
  - Although there is typically a small amount of extra latency to present the frame (and sometimes a large amount of extra latency), it is useful to know the effect of the main thread in isolation for several reasons:
  - As a performance metric, having both paint and presentation times available you can more easily know if the latency was caused by e.g. javascript or style/layout vs e.g. css effects or decode/rasterization work
  - Beyond that, the paintTime is very useful to segment the overall timeline.  E.g. if you want to understand what the state of the DOM was, and which tasks affected a long paint, you need to segment by the main thread time only.  Presentation time is an asynchronous value and any main thread updates that overlap with this latency don't directly affect the latency.


There is probably a lot more to be said, or some more presentations to share-- but the real TL;DR is:

- It helped make performance timeline interoperable (successfully so!)
- It has proven useful to developers even in implementations that have exposed both values together.

**Comment by xiaochengh** [comment]
@mmocny Thanks for the clarification!

The TAG discussed it today and are convinced of the value of `paintTime`. We would also like the spec to be clear about the difference between `paintTime` and `presentationTime`, especially in their intended use cases and limitations.

Therefore closing this review as `satisfied`.

**Comment by mmocny** [comment]
Thank you for the feedback and questions!
